File: typescript/core/test/rerank.test.ts
import { describe, expect, it } from 'vitest';
import { ChatClientLike, SmoothAgent } from '../src/agent.js';
import { InMemoryKnowledge, KnowledgeHit } from '../src/knowledge.js';
import { LexicalReranker, NoopReranker } from '../src/rerank.js';

const hit = (content: string, source = 's'): KnowledgeHit => ({ content, source, score: 0 });

describe('reranking', () => {
    it('NoopReranker is passthrough', () => {
        const hits = [hit('a'), hit('b')];
        expect(new NoopReranker().rerank('q', hits)).toEqual(hits);
    });

    it('LexicalReranker prefers a concise doc over a long one with same coverage', () => {
        const concise = hit('return policy');
        const verbose = hit(`return ${'filler '.repeat(60)}policy`);
        const out = new LexicalReranker().rerank('return policy', [verbose, concise]);
        expect(out[0]).toBe(concise);
        expect(out[1]).toBe(verbose);
    });

    it('LexicalReranker prefers higher coverage', () => {
        const coversTwo = hit('return and policy details');
        const coversOne = hit('return shipping details');
        const out = new LexicalReranker().rerank('return policy', [coversOne, coversTwo]);
        expect(out[0]).toBe(coversTwo);
    });

    it('empty query is passthrough', () => {
        const hits = [hit('a'), hit('b')];
        expect(new LexicalReranker().rerank('', hits)).toEqual(hits);
    });

    it('is applied between retrieval and injection', async () => {
        const kb = new InMemoryKnowledge();
        kb.ingest(`return ${'filler '.repeat(60)}policy`, 'long.md');
        kb.ingest('return policy', 'short.md');

        let systemContent = '';
        const client: ChatClientLike = {
            chat: {
                completions: {
                    create: async (body: Record<string, unknown>) => {
                        const messages = body.messages as Array<Record<string, unknown>>;
                        systemContent = (messages[0].content as string) ?? '';
                        return { choices: [{ message: { content: 'ok' } }] };
                    },
                },
            },
        };
        const agent = new SmoothAgent(client, {
            instructions: 'support',
            knowledge: kb,
            knowledgeTopK: 1,
            knowledgeCandidateK: 2,
            reranker: new LexicalReranker(),
        });
        await agent.run('return policy');
        expect(systemContent).toContain('[short.md]');
        expect(systemContent).not.toContain('[long.md]');
    });
});

File: go/core/rerank_test.go
package core

import (
	"strings"
	"testing"
)

func hit(content string) KnowledgeHit {
	return KnowledgeHit{Content: content, Source: "s", Score: 0}
}

func TestNoopRerankerPassthrough(t *testing.T) {
	hits := []KnowledgeHit{hit("a"), hit("b")}
	out := NoopReranker{}.Rerank("q", hits)
	if len(out) != 2 || out[0].Content != "a" {
		t.Fatalf("noop should be passthrough: %+v", out)
	}
}

func TestLexicalRerankerPrefersConcise(t *testing.T) {
	concise := hit("return policy")
	verbose := hit("return " + strings.Repeat("filler ", 60) + "policy")
	out := LexicalReranker{}.Rerank("return policy", []KnowledgeHit{verbose, concise})
	if out[0].Content != concise.Content {
		t.Fatalf("concise doc should rank first: %+v", out)
	}
}

func TestLexicalRerankerPrefersHigherCoverage(t *testing.T) {
	coversTwo := hit("return and policy details")
	coversOne := hit("return shipping details")
	out := LexicalReranker{}.Rerank("return policy", []KnowledgeHit{coversOne, coversTwo})
	if out[0].Content != coversTwo.Content {
		t.Fatalf("higher-coverage doc should rank first: %+v", out)
	}
}

func TestLexicalRerankerEmptyQueryPassthrough(t *testing.T) {
	hits := []KnowledgeHit{hit("a"), hit("b")}
	out := LexicalReranker{}.Rerank("", hits)
	if out[0].Content != "a" {
		t.Fatalf("empty query should be passthrough")
	}
}

func TestRerankerAppliedInBuildSystem(t *testing.T) {
	kb := &InMemoryKnowledge{}
	kb.Ingest("return "+strings.Repeat("filler ", 60)+"policy", "long.md")
	kb.Ingest("return policy", "short.md")

	agent := NewSmoothAgent(&fakeClient{}, AgentOptions{
		Instructions:        "support",
		Knowledge:           kb,
		KnowledgeTopK:       1,
		KnowledgeCandidateK: 2,
		Reranker:            LexicalReranker{},
	})
	system := agent.buildSystem("return policy")
	if !strings.Contains(system, "[short.md]") {
		t.Fatalf("reranked top-1 (short.md) should be injected: %q", system)
	}
	if strings.Contains(system, "[long.md]") {
		t.Fatalf("long.md should have been reranked out: %q", system)
	}
}
